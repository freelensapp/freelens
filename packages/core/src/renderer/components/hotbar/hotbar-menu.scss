/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "../vars" as *;

.HotbarMenu {
  $spacing: $padding * 2;

  position: relative;
  text-align: center;
  background: var(--clusterMenuBackground);
  padding-top: 1px;
  width: var(--hotbar-width);
  overflow: hidden;

  &.draggingOver::after {
    content: " ";
    position: fixed;
    left: var(--hotbar-width);
    width: 100%;
    height: 100%;
  }

  .HotbarItems {
    --cellWidth: 40px;
    --cellHeight: 40px;

    box-sizing: content-box;
    margin: 0 auto;
    height: 100%;
    overflow: hidden;
    padding-bottom: 8px;

    &:hover {
      overflow: overlay;

      &::-webkit-scrollbar {
        width: 0.4em;
        background: transparent;
        z-index: 1;
      }

      &::-webkit-scrollbar-thumb {
        background: var(--borderFaintColor);
      }
    }

    .HotbarCell {
      width: var(--cellWidth);
      height: var(--cellHeight);
      min-height: var(--cellHeight);
      margin: 12px;
      background: var(--clusterMenuCellBackground);
      border-radius: 6px;
      position: relative;

      &.isDraggingOver {
        background-color: #3e4148;
        box-shadow: 0 0 0px 3px var(--clusterMenuBackground), 0 0 0px 6px #ffffff30;

        &:not(.isDraggingOwner) {
          z-index: 50;

          > div:not(:empty) {
            border-radius: 6px;
            box-shadow: 0 0 9px #00000042;
          }

          &.animateUp {
            > div {
              transform: translate(0px, -40px)!important;
            }
          }

          &.animateDown {
            > div {
              transform: translate(0px, 40px);
            }
          }
        }
      }

      &.animating {
        &:empty {
          animation: shake .6s cubic-bezier(.36,.07,.19,.97) both;
          transform: translate3d(0, 0, 0);
          backface-visibility: hidden;
          perspective: 1000px;
        }

        &:not(:empty) {
          > div {
            animation: click .1s;
          }
        }
      }
    }
  }
}

@keyframes shake {
  10%, 90% {
    transform: translate3d(-1px, 0, 0);
  }

  20%, 80% {
    transform: translate3d(2px, 0, 0);
  }

  30%, 50%, 70% {
    transform: translate3d(-4px, 0, 0);
  }

  40%, 60% {
    transform: translate3d(4px, 0, 0);
  }
}

@keyframes click {
  0% {
    margin-top: 0;
  }

  100% {
    margin-top: 2px;
  }
}
