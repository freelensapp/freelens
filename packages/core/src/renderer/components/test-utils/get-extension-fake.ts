/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { LensMainExtension } from "../../../extensions/lens-main-extension";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

export interface FakeExtensionOptions {
  id: string;
  name: string;
  rendererOptions?: Partial<LensRendererExtension>;
  mainOptions?: Partial<LensMainExtension>;
}

export const getExtensionFakeForMain = ({ id, name, mainOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionMain({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    mainOptions,
  );

export const getExtensionFakeForRenderer = ({ id, name, rendererOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionRenderer({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    rendererOptions,
  );
