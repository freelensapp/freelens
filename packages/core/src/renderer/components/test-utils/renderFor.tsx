/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
