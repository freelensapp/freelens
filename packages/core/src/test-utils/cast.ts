/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

export const cast = <T>(data: Partial<T>): T => data as T;
