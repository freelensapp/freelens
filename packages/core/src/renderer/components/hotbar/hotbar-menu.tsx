/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./hotbar-menu.scss";

import type { IClassName, StrictReactNode } from "@freelensapp/utilities";
import { cssNames } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import type { IComputedValue } from "mobx";
import { observer } from "mobx-react";
import React, { useState } from "react";
import { DragDropContext, Draggable, type DropResult, Droppable } from "react-beautiful-dnd";
import activeHotbarInjectable from "../../../features/hotbar/storage/common/active.injectable";
import type { Hotbar } from "../../../features/hotbar/storage/common/hotbar";
import type { HotbarItem } from "../../../features/hotbar/storage/common/types";
import { defaultHotbarCells } from "../../../features/hotbar/storage/common/types";
import type { CatalogEntity } from "../../api/catalog-entity";
import type { CatalogEntityRegistry } from "../../api/catalog/entity/registry";
import catalogEntityRegistryInjectable from "../../api/catalog/entity/registry.injectable";
import { HotbarCell } from "./hotbar-cell";
import { HotbarEntityIcon } from "./hotbar-entity-icon";
import { HotbarIcon } from "./hotbar-icon";
import { HotbarSelector } from "./hotbar-selector";

export interface HotbarMenuProps {
  className?: IClassName;
}

interface Dependencies {
  activeHotbar: IComputedValue<Hotbar | undefined>;
  entityRegistry: CatalogEntityRegistry;
}

const NonInjectedHotbarMenu = observer((props: Dependencies & HotbarMenuProps) => {
  const { activeHotbar, entityRegistry, className } = props;

  const [draggingOver, setDraggingOver] = useState(false);
  const hotbar = activeHotbar.get();

  const getEntity = (item: HotbarItem | null) => {
    if (!item) {
      return undefined;
    }

    return entityRegistry.getById(item.entity.uid);
  };
  const onDragStart = () => setDraggingOver(true);
  const onDragEnd = (result: DropResult) => {
    setDraggingOver(false);

    const { source, destination } = result;

    if (!destination) {
      // Dropped outside of the list
      return;
    }

    const from = parseInt(source.droppableId);
    const to = parseInt(destination.droppableId);

    hotbar?.restack(from, to);
  };
  const removeItem = (entityId: string) => {
    hotbar?.removeEntity(entityId);
  };
  const addItem = (entity: CatalogEntity) => {
    hotbar?.addEntity(entity);
  };
  const getMoveAwayDirection = (entityId: string | undefined | null, cellIndex: number) => {
    if (!entityId || !hotbar) {
      return "animateDown";
    }

    const draggableItemIndex = hotbar.items.findIndex((item) => item?.entity.uid == entityId);

    return draggableItemIndex > cellIndex ? "animateDown" : "animateUp";
  };

  const renderGrid = () =>
    hotbar?.items.map((item, index) => {
      const entity = getEntity(item);

      return (
        <Droppable droppableId={`${index}`} key={index}>
          {(provided, snapshot) => (
            <HotbarCell
              index={index}
              key={entity ? entity.getId() : `cell${index}`}
              innerRef={provided.innerRef}
              className={cssNames(
                {
                  isDraggingOver: snapshot.isDraggingOver,
                  isDraggingOwner: snapshot.draggingOverWith == entity?.getId(),
                },
                getMoveAwayDirection(snapshot.draggingOverWith, index),
              )}
              {...provided.droppableProps}
            >
              {item && (
                <Draggable draggableId={item.entity.uid} key={item.entity.uid} index={0}>
                  {(provided, snapshot) => (
                    <div
                      key={item.entity.uid}
                      ref={provided.innerRef}
                      {...provided.draggableProps}
                      {...provided.dragHandleProps}
                      style={{
                        zIndex: defaultHotbarCells - index,
                        position: "absolute" as const,
                        ...provided.draggableProps.style,
                      }}
                    >
                      {entity ? (
                        <HotbarEntityIcon
                          key={index}
                          index={index}
                          entity={entity}
                          onClick={() => entityRegistry.onRun(entity)}
                          className={cssNames({ isDragging: snapshot.isDragging })}
                          remove={removeItem}
                          add={addItem}
                          size={40}
                        />
                      ) : (
                        <HotbarIcon
                          uid={`hotbar-icon-${item.entity.uid}`}
                          title={item.entity.name}
                          source={item.entity.source ?? "local"}
                          tooltip={`${item.entity.name} (${item.entity.source})`}
                          menuItems={[
                            {
                              title: "Remove from Hotbar",
                              onClick: () => removeItem(item.entity.uid),
                            },
                          ]}
                          disabled
                          size={40}
                        />
                      )}
                    </div>
                  )}
                </Draggable>
              )}
              {provided.placeholder as StrictReactNode}
            </HotbarCell>
          )}
        </Droppable>
      );
    });

  return (
    <div className={cssNames("HotbarMenu flex column", { draggingOver }, className)}>
      <div className="HotbarItems flex column gaps">
        <DragDropContext onDragStart={() => onDragStart()} onDragEnd={(result) => onDragEnd(result)}>
          {renderGrid()}
        </DragDropContext>
      </div>
      <HotbarSelector />
    </div>
  );
});

export const HotbarMenu = withInjectables<Dependencies, HotbarMenuProps>(NonInjectedHotbarMenu, {
  getProps: (di, props) => ({
    ...props,
    entityRegistry: di.inject(catalogEntityRegistryInjectable),
    activeHotbar: di.inject(activeHotbarInjectable),
  }),
});
