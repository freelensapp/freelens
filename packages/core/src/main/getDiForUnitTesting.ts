/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { registerFeature } from "@freelensapp/feature-core";
import { kubeApiSpecificsFeature } from "@freelensapp/kube-api-specifics";
import { setLegacyGlobalDiForExtensionApi } from "@freelensapp/legacy-global-di";
import { loggerFeature } from "@freelensapp/logger";
import { messagingFeature, testUtils as messagingTestUtils } from "@freelensapp/messaging";
import { notificationsFeature } from "@freelensapp/notifications";
import { randomFeature } from "@freelensapp/random";
import type { GlobalOverride } from "@freelensapp/test-utils";
import type { DiContainer } from "@ogre-tools/injectable";
import { createContainer, isInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { chunk } from "lodash/fp";
import { runInAction } from "mobx";
import broadcastMessageInjectable from "../common/ipc/broadcast-message.injectable";
import setupSyncingOfWeblinksInjectable from "../features/weblinks/main/setup-syncing-of-weblinks.injectable";
import { getOverrideFsWithFakes } from "../test-utils/override-fs-with-fakes";
import spawnInjectable from "./child-process/spawn.injectable";
import initializeClusterManagerInjectable from "./cluster/initialize-manager.injectable";
import electronQuitAndInstallUpdateInjectable from "./electron-app/features/electron-quit-and-install-update.injectable";
import electronUpdaterIsActiveInjectable from "./electron-app/features/electron-updater-is-active.injectable";
import setUpdateOnQuitInjectable from "./electron-app/features/set-update-on-quit.injectable";
import setupApplicationNameInjectable from "./electron-app/runnables/setup-application-name.injectable";
import setupDeepLinkingInjectable from "./electron-app/runnables/setup-deep-linking.injectable";
import setupDeviceShutdownInjectable from "./electron-app/runnables/setup-device-shutdown.injectable";
import setupIpcMainHandlersInjectable from "./electron-app/runnables/setup-ipc-main-handlers/setup-ipc-main-handlers.injectable";
import setupMainWindowVisibilityAfterActivationInjectable from "./electron-app/runnables/setup-main-window-visibility-after-activation.injectable";
import waitUntilBundledExtensionsAreLoadedInjectable from "./start-main-application/lens-window/application-window/wait-until-bundled-extensions-are-loaded.injectable";
import initializeExtensionsInjectable from "./start-main-application/runnables/initialize-extensions.injectable";
import setupLensProxyInjectable from "./start-main-application/runnables/setup-lens-proxy.injectable";

export function getDiForUnitTesting() {
  const environment = "main";
  const di = createContainer(environment, {
    detectCycles: false,
  });

  registerMobX(di);
  setLegacyGlobalDiForExtensionApi(di, environment);

  runInAction(() => {
    registerFeature(
      di,
      messagingFeature,
      messagingTestUtils.messagingFeatureForUnitTesting,
      loggerFeature,
      randomFeature,
      kubeApiSpecificsFeature,
      notificationsFeature,
    );
  });

  di.preventSideEffects();

  runInAction(() => {
    const injectables = global.injectablePaths.main.paths
      .map((path) => require(path))
      .flatMap(Object.values)
      .filter(isInjectable);

    for (const block of chunk(100)(injectables)) {
      di.register(...block);
    }
  });

  for (const globalOverridePath of global.injectablePaths.main.globalOverridePaths) {
    const globalOverride = require(globalOverridePath).default as GlobalOverride<unknown, unknown, unknown>;

    di.override(globalOverride.injectable, globalOverride.overridingInstantiate);
  }

  di.override(waitUntilBundledExtensionsAreLoadedInjectable, () => async () => {});

  overrideRunnablesHavingSideEffects(di);
  overrideElectronFeatures(di);
  getOverrideFsWithFakes()(di);

  di.override(broadcastMessageInjectable, () => (channel) => {
    throw new Error(`Tried to broadcast message to channel "${channel}" over IPC without explicit override.`);
  });
  di.override(spawnInjectable, () => () => {
    return {
      stderr: { on: jest.fn(), removeAllListeners: jest.fn() },
      stdout: { on: jest.fn(), removeAllListeners: jest.fn() },
      on: jest.fn(),
    } as never;
  });

  return di;
}

// TODO: Reorganize code in Runnables to get rid of requirement for override
const overrideRunnablesHavingSideEffects = (di: DiContainer) => {
  [
    initializeExtensionsInjectable,
    initializeClusterManagerInjectable,
    setupIpcMainHandlersInjectable,
    setupLensProxyInjectable,
    setupSyncingOfWeblinksInjectable,
  ].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });
};

const overrideElectronFeatures = (di: DiContainer) => {
  [
    setupMainWindowVisibilityAfterActivationInjectable,
    setupDeviceShutdownInjectable,
    setupDeepLinkingInjectable,
    setupApplicationNameInjectable,
  ].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });

  di.override(electronQuitAndInstallUpdateInjectable, () => () => {});
  di.override(setUpdateOnQuitInjectable, () => () => {});
  di.override(electronUpdaterIsActiveInjectable, () => false);
};
