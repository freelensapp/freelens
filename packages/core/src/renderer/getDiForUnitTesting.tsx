/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { animateFeature, requestAnimationFrameInjectable } from "@freelensapp/animate";
import { clusterSidebarFeature } from "@freelensapp/cluster-sidebar";
import { registerFeature } from "@freelensapp/feature-core";
import { kubeApiSpecificsFeature } from "@freelensapp/kube-api-specifics";
import { setLegacyGlobalDiForExtensionApi } from "@freelensapp/legacy-global-di";
import { loggerFeature } from "@freelensapp/logger";
import { messagingFeature, testUtils as messagingTestUtils } from "@freelensapp/messaging";
import { notificationsFeature } from "@freelensapp/notifications";
import { randomFeature } from "@freelensapp/random";
import { routingFeature } from "@freelensapp/routing";
import type { GlobalOverride } from "@freelensapp/test-utils";
import { createContainer, isInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { registerInjectableReact } from "@ogre-tools/injectable-react";
import { chunk, noop } from "lodash/fp";
import { runInAction } from "mobx";
import { getOverrideFsWithFakes } from "../test-utils/override-fs-with-fakes";
import hostedClusterIdInjectable from "./cluster-frame-context/hosted-cluster-id.injectable";
import terminalSpawningPoolInjectable from "./components/dock/terminal/terminal-spawning-pool.injectable";
import startTopbarStateSyncInjectable from "./components/layout/top-bar/start-state-sync.injectable";
import legacyOnChannelListenInjectable from "./ipc/legacy-channel-listen.injectable";
import watchHistoryStateInjectable from "./remote-helpers/watch-history-state.injectable";

export const getDiForUnitTesting = () => {
  const environment = "renderer";
  const di = createContainer(environment, {
    detectCycles: false,
  });

  registerMobX(di);
  registerInjectableReact(di);
  setLegacyGlobalDiForExtensionApi(di, environment);

  runInAction(() => {
    registerFeature(
      di,
      messagingFeature,
      messagingTestUtils.messagingFeatureForUnitTesting,
      routingFeature,
      loggerFeature,
      animateFeature,
      clusterSidebarFeature,
      randomFeature,
      kubeApiSpecificsFeature,
      notificationsFeature,
    );
  });

  di.preventSideEffects();

  runInAction(() => {
    const injectables = global.injectablePaths.renderer.paths
      .map((path) => require(path))
      .flatMap(Object.values)
      .filter(isInjectable);

    for (const block of chunk(100)(injectables)) {
      di.register(...block);
    }
  });

  for (const globalOverridePath of global.injectablePaths.renderer.globalOverridePaths) {
    const globalOverride = require(globalOverridePath).default as GlobalOverride<unknown, unknown, unknown>;

    di.override(globalOverride.injectable, globalOverride.overridingInstantiate);
  }

  [startTopbarStateSyncInjectable].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });

  di.override(terminalSpawningPoolInjectable, () => document.createElement("div"));
  di.override(hostedClusterIdInjectable, () => undefined);

  di.override(legacyOnChannelListenInjectable, () => () => noop);

  di.override(requestAnimationFrameInjectable, () => (callback) => callback());
  di.override(watchHistoryStateInjectable, () => () => () => {});

  getOverrideFsWithFakes()(di);

  return di;
};
