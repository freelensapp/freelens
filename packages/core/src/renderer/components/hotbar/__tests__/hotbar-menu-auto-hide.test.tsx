/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "@testing-library/jest-dom";

import { fireEvent, waitFor } from "@testing-library/react";
import React from "react";
import directoryForUserDataInjectable from "../../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import userPreferencesStateInjectable from "../../../../features/user-preferences/common/state.injectable";
import { getDiForUnitTesting } from "../../../getDiForUnitTesting";
import { renderFor } from "../../test-utils/renderFor";
import { HotbarMenu } from "../hotbar-menu";

import type { RenderResult } from "@testing-library/react";

import type { UserPreferencesState } from "../../../../features/user-preferences/common/state.injectable";

describe("<HotbarMenu /> auto-hide functionality", () => {
  let result: RenderResult;
  let userPreferencesState: UserPreferencesState;

  beforeEach(() => {
    const di = getDiForUnitTesting();
    const render = renderFor(di);

    di.override(directoryForUserDataInjectable, () => "/some-directory-for-user-data");

    userPreferencesState = di.inject(userPreferencesStateInjectable);

    result = render(<HotbarMenu />);
  });

  describe("when auto-hide is disabled", () => {
    beforeEach(() => {
      userPreferencesState.hotbarAutoHide = false;
      result.rerender(<HotbarMenu />);
    });

    it("renders w/o errors", () => {
      expect(result.container).toMatchSnapshot();
    });

    it("should not have autoHide class", () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");
      expect(hotbarMenu).not.toHaveClass("autoHide");
    });

    it("should not have visible class", () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");
      expect(hotbarMenu).not.toHaveClass("visible");
    });
  });

  describe("when auto-hide is enabled", () => {
    beforeEach(() => {
      userPreferencesState.hotbarAutoHide = true;
      result.rerender(<HotbarMenu />);
    });

    it("renders w/o errors", () => {
      expect(result.container).toMatchSnapshot();
    });

    it("should have autoHide class", () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");
      expect(hotbarMenu).toHaveClass("autoHide");
    });

    it("should start hidden (without visible class)", () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");
      expect(hotbarMenu).not.toHaveClass("visible");
    });

    it("should show hotbar when mouse moves to left edge", async () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");

      fireEvent.mouseMove(window, { clientX: 5 });

      await waitFor(() => {
        expect(hotbarMenu).toHaveClass("visible");
      });
    });

    it("should hide hotbar when mouse moves away", async () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");

      // Show first
      fireEvent.mouseMove(window, { clientX: 5 });
      await waitFor(() => expect(hotbarMenu).toHaveClass("visible"));

      // Then hide
      fireEvent.mouseMove(window, { clientX: 100 });

      await waitFor(() => {
        expect(hotbarMenu).not.toHaveClass("visible");
      });
    });

    it("should hide hotbar on mouseLeave event", async () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");

      fireEvent.mouseMove(window, { clientX: 5 });
      await waitFor(() => expect(hotbarMenu).toHaveClass("visible"));

      if (hotbarMenu) {
        fireEvent.mouseLeave(hotbarMenu);
      }

      await waitFor(() => {
        expect(hotbarMenu).not.toHaveClass("visible");
      });
    });

    it("should cleanup event listeners on unmount", () => {
      const removeEventListenerSpy = jest.spyOn(window, "removeEventListener");

      result.unmount();

      expect(removeEventListenerSpy).toHaveBeenCalledWith("mousemove", expect.any(Function));

      removeEventListenerSpy.mockRestore();
    });
  });

  describe("trigger zone interaction", () => {
    let triggerZone: HTMLElement;

    beforeEach(() => {
      userPreferencesState.hotbarAutoHide = true;

      triggerZone = document.createElement("div");
      triggerZone.className = "hotbar-trigger-zone";
      document.body.appendChild(triggerZone);

      result.rerender(<HotbarMenu />);
    });

    afterEach(() => {
      if (triggerZone && triggerZone.parentNode) {
        triggerZone.parentNode.removeChild(triggerZone);
      }
    });

    it("should show hotbar when mouse enters trigger zone", async () => {
      const hotbarMenu = result.container.querySelector(".HotbarMenu");

      fireEvent.mouseEnter(triggerZone);

      await waitFor(() => {
        expect(hotbarMenu).toHaveClass("visible");
      });
    });
  });
});
