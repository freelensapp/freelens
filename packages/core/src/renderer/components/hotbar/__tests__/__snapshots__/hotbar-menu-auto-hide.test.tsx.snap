// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HotbarMenu /> auto-hide functionality when auto-hide is disabled renders w/o errors 1`] = `
<div>
  <div
    class="HotbarMenu flex column"
  >
    <div
      class="HotbarItems flex column gaps"
    />
    <div
      class="HotbarSelector"
    >
      <i
        class="Icon material interactive focusable"
        tabindex="0"
      >
        <span
          class="icon"
          data-icon-name="arrow_left"
        >
          arrow_left
        </span>
      </i>
      <div
        class="HotbarIndex"
      >
        <div
          class="badge Badge small clickable"
          data-testid="hotbar-menu-badge-for-undefined"
          id="hotbarIndex"
        >
          ??
        </div>
      </div>
      <i
        class="Icon material interactive focusable"
        tabindex="0"
      >
        <span
          class="icon"
          data-icon-name="arrow_right"
        >
          arrow_right
        </span>
      </i>
    </div>
  </div>
</div>
`;

exports[`<HotbarMenu /> auto-hide functionality when auto-hide is enabled renders w/o errors 1`] = `
<div>
  <div
    class="HotbarMenu flex column autoHide"
  >
    <div
      class="HotbarItems flex column gaps"
    />
    <div
      class="HotbarSelector"
    >
      <i
        class="Icon material interactive focusable"
        tabindex="0"
      >
        <span
          class="icon"
          data-icon-name="arrow_left"
        >
          arrow_left
        </span>
      </i>
      <div
        class="HotbarIndex"
      >
        <div
          class="badge Badge small clickable"
          data-testid="hotbar-menu-badge-for-undefined"
          id="hotbarIndex"
        >
          ??
        </div>
      </div>
      <i
        class="Icon material interactive focusable"
        tabindex="0"
      >
        <span
          class="icon"
          data-icon-name="arrow_right"
        >
          arrow_right
        </span>
      </i>
    </div>
  </div>
</div>
`;
